scalding-core/src/main/scala/com/twitter/scalding/Operations.scala

```diff
@@ -166,30 +156,93 @@ package com.twitter.scalding {
       }
     }
 
-    override def operate(flowProcess: FlowProcess[_], functionCall: FunctionCall[SummingCache[Tuple, V]]) {
+    override def operate(flowProcess: FlowProcess[_], functionCall: FunctionCall[MapsideCache[V]]) {
       val cache = functionCall.getContext
       val keyValueTE = functionCall.getArguments
       // Have to keep a copy of the key tuple because cascading will modify it
       val key = keyValueTE.selectEntry(keyFields).getTupleCopy
       val value = conv(keyValueTE.selectEntry(valueFields))
-      add(cache.put(Map(key -> value)), functionCall)
+      val (stats, evicted) = cache.put(key, value)
+      add(evicted, functionCall)
+
+      //iterator for performance
+      val it = stats.iterator
+      while (it.hasNext) {
+        val (key, value) = it.next
+        flowProcess.increment(COUNTER_GROUP, key, value)
+      }
+
+      if (evicted.isDefined)
+        flowProcess.increment(COUNTER_GROUP, "evictions", evicted.get.size)
     }
 
-    override def flush(flowProcess: FlowProcess[_], operationCall: OperationCall[SummingCache[Tuple, V]]) {
+    override def flush(flowProcess: FlowProcess[_], operationCall: OperationCall[MapsideCache[V]]) {
       // Docs say it is safe to do this cast:
       // http://docs.cascading.org/cascading/2.1/javadoc/cascading/operation/Operation.html#flush(cascading.flow.FlowProcess, cascading.operation.OperationCall)
-      val functionCall = operationCall.asInstanceOf[FunctionCall[SummingCache[Tuple, V]]]
+      val functionCall = operationCall.asInstanceOf[FunctionCall[MapsideCache[V]]]
       val cache = functionCall.getContext
       add(cache.flush, functionCall)
     }
 
-    override def cleanup(flowProcess: FlowProcess[_], operationCall: OperationCall[SummingCache[Tuple, V]]) {
+    override def cleanup(flowProcess: FlowProcess[_], operationCall: OperationCall[MapsideCache[V]]) {
       // The cache may be large, but super sure we drop any reference to it ASAP
       // probably overly defensive, but it's super cheap.
       operationCall.setContext(null)
     }
   }
 
+  sealed trait MapsideCache[V] {
+    def flush: Option[Map[Tuple, V]]
+    def put(key: Tuple, value: V): (Map[String, Int], Option[Map[Tuple, V]])
+  }
+
+  object MapsideCache {
+    val DEFAULT_CACHE_SIZE = 100000
+    val SIZE_CONFIG_KEY = AggregateBy.AGGREGATE_BY_THRESHOLD
+    val ADAPTIVE_CACHE_KEY = "scalding.mapsidecache.adaptive"
+
+    private def getCacheSize(fp: FlowProcess[_]): Int =
+      Option(fp.getStringProperty(SIZE_CONFIG_KEY))
+        .filterNot { _.isEmpty }
+        .map { _.toInt }
+        .getOrElse(DEFAULT_CACHE_SIZE)
+
+    def apply[V: Semigroup](cacheSize: Option[Int], flowProcess: FlowProcess[_]): MapsideCache[V] = {
+      val size = cacheSize.getOrElse{ getCacheSize(flowProcess) }
+      val adaptive = Option(flowProcess.getStringProperty(ADAPTIVE_CACHE_KEY)).isDefined
```

Comment:
What if this is an empty string? (Presumably that's possible, since we check for that with `fp.getStringProperty(SIZE_CONFIG_KEY)` on line 206 above.) Or if someone gets creative and sets it explicitly to false?

I'd think it should be something like `val adaptive = Option(flowProcess.getStringProperty(ADAPTIVE_CACHE_KEY)) == Some("true")` instead.